consensus/src/network_interface.rs

```diff
@@ -139,9 +164,41 @@ impl ConsensusNetworkSender {
         message: ConsensusMsg,
         timeout: Duration,
     ) -> Result<ConsensusMsg, RpcError> {
-        let protocol = ProtocolId::ConsensusRpc;
+        let protocol = self.preferred_protocol_for_peer(recipient, RPC)?;
         self.network_sender
             .send_rpc(recipient, protocol, message, timeout)
             .await
     }
+
+    /// Initialize a shared hashmap about connections metadata that is updated by the receiver.
+    pub fn initialize(&mut self, connections: Arc<RwLock<HashMap<PeerId, SupportedProtocols>>>) {
+        self.peers_protocols = Some(connections);
+    }
+
+    /// Query the supported protocols from this peer's connection.
+    fn supported_protocols(&self, peer: PeerId) -> anyhow::Result<SupportedProtocols> {
+        if let Some(map) = &self.peers_protocols {
+            map.read()
+                .get(&peer)
+                .cloned()
+                .ok_or_else(|| anyhow!("Peer not connected"))
+        } else {
+            Err(anyhow!("ConsensusNetworkSender not initialized"))
+        }
+    }
+
+    /// Choose the overlapping protocol for peer.
```

Comment:
Seems a comment is missing here where the local_protocols are sorted from most to least preferred.